tower-http/src/classify.rs

```diff
@@ -0,0 +1,170 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Response, StatusCode};
+use http_body::Body;
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for classifying responses as either success or failure.
+pub trait ClassifyResponse {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// The type used to classify the end of streaming responses.
+    type ClassifyEos: ClassifyEos<Output = Self::Output>;
```

Comment:
See my other comment, but I'd maybe define "Eos" in the documentation:

```suggestion
    /// The type used to classify the end of streaming (Eos) responses.
    type ClassifyEos: ClassifyEos<Output = Self::Output>;
```